swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,928 @@
+    where
+        TOutboundStream: Stream<Item = TOutbound> + Send + 'static,
+    {
+        FromFnProto {
+            protocol: self.protocol,
+            on_new_inbound: self.phase.inbound_handler,
+            on_new_outbound: Box::new(move |stream, peer, endpoint, state, info| {
+                handler(stream, peer, endpoint, state, info).boxed()
+            }),
+            inbound_streams_limit: self.phase.max_inbound,
+            pending_outbound_streams_limit: max_pending_outbound,
+            state: self.phase.state,
+        }
+    }
+
+    pub fn without_outbound_handler(self) -> FromFnProto<TInbound, Void, Void, TState> {
+        FromFnProto {
+            protocol: self.protocol,
+            on_new_inbound: self.phase.inbound_handler,
+            on_new_outbound: Box::new(|_, _, _, _, info| void::unreachable(info)),
+            inbound_streams_limit: self.phase.max_inbound,
+            pending_outbound_streams_limit: 0,
+            state: self.phase.state,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum OutEvent<I, O, OpenInfo> {
+    InboundEmitted(I),
+    OutboundEmitted(O),
+    FailedToOpen(OpenError<OpenInfo>),
+}
+
+#[derive(Debug)]
+pub enum OpenError<OpenInfo> {
+    /// The time limit for the negotiation handshake was exceeded.
+    Timeout(OpenInfo),
+    /// We have hit the configured limit for the maximum number of pending substreams.
+    LimitExceeded {
+        open_info: OpenInfo,
+        pending_substreams: usize,
+    },
+    /// The remote does not support this protocol.
+    Unsupported {
+        open_info: OpenInfo,
+        protocol: &'static str,
+    },
+    NegotiationFailed(OpenInfo, ProtocolError),
+}
+
+/// A wrapper for state that is shared across all connections.
+///
+/// Any update to the state will "automatically" be relayed to all connections, assuming this struct
+/// is correctly wired into your [`NetworkBehaviour`](crate::swarm::NetworkBehaviour).
+///
+/// This struct implements an observer pattern. All registered connections will receive updates that
+/// are made to the state.
+pub struct Shared<T> {
+    inner: T,
+
+    shared: Arc<T>,
+
+    dirty: bool,
+    waker: Option<Waker>,
+    connections: HashSet<(PeerId, ConnectionId)>,
+    pending_update_events: VecDeque<(PeerId, ConnectionId, Arc<T>)>,
+}
+
+impl<T> Shared<T>
+where
+    T: Clone,
+{
+    pub fn new(state: T) -> Self {
+        Self {
+            inner: state.clone(),
+            shared: Arc::new(state),
+            dirty: false,
+            waker: None,
+            connections: HashSet::default(),
+            pending_update_events: VecDeque::default(),
+        }
+    }
+
+    pub fn register_connection(&mut self, peer_id: PeerId, id: ConnectionId) {
+        self.connections.insert((peer_id, id));
+    }
+
+    pub fn unregister_connection(&mut self, peer_id: PeerId, id: ConnectionId) {
+        self.connections.remove(&(peer_id, id));
+    }
```

Comment:
Once https://github.com/libp2p/rust-libp2p/issues/2832 is fixed, we can make this more ergonomic by just requiring a reference to the `FromSwarm` event and we can pick out the data as we need.